### ISSUE TITLE
Getting WASM build to work for custom CSS plugin

### ISSUE THREAD

**Issue by disrupted** [dprint/dprint#447]
Hi David, not sure if this is the right place to ask.

A few days ago the new [Parcel CSS](https://github.com/parcel-bundler/parcel-css) crate was released. Since dprint doesn't have a native CSS plugin yet, I thought of looking into it and trying to set up a very simple plugin. (just using the default parcel CSS formatting).

I followed the [instructions for creating a WASM plugin](https://github.com/dprint/dprint/blob/main/docs/wasm-plugin-development.md) and here's what I have so far: https://github.com/disrupted/dprint-plugin-css

After building it using
```sh
cargo build --target wasm32-unknown-unknown --release --features "wasm"
```
I run it with
```sh
dprint fmt --plugins target/wasm32-unknown-unknown/release/dprint_plugin_css.wasm -- style.css
```

For now this doesn't work because of the following error:
```
Error resolving plugin /dev/dprint-plugin-css/target/wasm32-unknown-unknown/release/dprint_plugin_css.wasm: 
Error instantiating module: Error while importing "__wbindgen_placeholder__"."__wbindgen_describe": unknown import. 
Expected Function(FunctionType { params: [I32], results: [] })
```

My guess is, this has something to do with how the parcel_css module is bundled during the WASM build. Perhaps you can tell, if that's something common or if it's something that needs to be fixed within parcel_css.

btw once this works, I could very well imagine this being turned into an actual formatting plugin, since parcel already provides an AST.

**Comment by dsherret** [comment]
@disrupted Oh neat! Glad you're working on that 😃

> My guess is, this has something to do with how the parcel_css module is bundled during the WASM build. Perhaps you can tell, if that's something common or if it's something that needs to be fixed within parcel_css.

Yes, it looks like something is adding in extra wasm exports using wasm-bindgen when doing a wasm build. Ideally all crates should have this off by default instead of creating rogue wasm exports.

Streaming investigation notes: Looking at the Cargo.lock file in the repo you sent and searching for "wasm-bindgen" it looks like that is used by parcel_sourcemap. parcel_sourcemap has a "wasm" feature that is not enabled by default though and wasm-bindgen is under that. Looking downstream to parcel_css, which has a dependency on parcel_sourcemap, it does not use the "wasm" feature..... hmmm... well actually parcel_sourcemap was published as 2.0 four months ago and the change was only recently made 15 days ago to create the "wasm" feature as should be done:

https://github.com/parcel-bundler/source-map/commit/99c1c9caeed2117e9a036edd419aa5c2b3d85686#diff-ee7df18dba7754a8793fd43cae48b9e93236e16c34d9d52092e1c4f5787ca79e

So, basically there just needs to be a new release done for parcel_sourcemap because it's currently pulling in wasm-bindgen by default, which is adding exports to allow its api to be used in wasm (which doesn't make sense for mostly any crate using it from rust in wasm).

**Comment by disrupted** [comment]
@dsherret thanks for your super quick reply and helping me to investigate this error! 🙂 After checking your link, I saw that there is already an open issue for the missing release: https://github.com/parcel-bundler/source-map/issues/89. I am gonna wait for this to be resolved and come back to the plugin afterwards.

**Comment by disrupted** [comment]
@dsherret the issue was fixed in the meantime and I updated the dependency. Now the formatter is working. 🙂 
Now I should figure out how to implement rules & generate the IR to turn it into a full featured dprint plugin. I noticed the [docs](https://github.com/dprint/dprint/blob/main/docs/overview.md#overview) are saying this part is outdated, but I guess I can compare it to existing plugins to figure out how this works.
thanks again for your help!
